doc/nvgre_tunnel/nvgre_tunnel.md

```diff
@@ -0,0 +1,437 @@
+<img src="images/nvgre_tunnel_delete_uml.svg" alt="Figure 4. NVGRE Tunnel delete flow">
+</p>
+
+##### Figure 5. NVGRE Tunnel Map delete flow
+
+<p align=center>
+<img src="images/nvgre_tunnel_map_delete_uml.svg" alt="Figure 5. NVGRE Tunnel Map delete flow">
+</p>
+
+##### Figure 6. NVGRE Tunnel CLI config
+
+<p align=center>
+<img src="images/nvgre_cli_config.svg" alt="Figure 6. NVGRE Tunnel CLI config">
+</p>
+
+##### Figure 7. NVGRE Tunnel CLI show
+
+<p align=center>
+<img src="images/nvgre_cli_show.svg" alt="Figure 7. NVGRE Tunnel CLI show">
+</p>
+
+### SAI API 
+
+The NVGRE Tunnel feature require at least `SAI 1.9` or above.
+
+| NVGRE component | SAI attribute |
+|--------------|---------------------------------------|
+| NVGRE tunnel type | SAI_TUNNEL_TYPE_NVGRE |
+| Encap mapper | SAI_TUNNEL_MAP_TYPE_VLAN_ID_TO_VSID |
+| Decap mapper | SAI_TUNNEL_MAP_TYPE_VSID_TO_VLAN_ID |
+| Encap mapper | SAI_TUNNEL_MAP_TYPE_BRIDGE_IF_TO_VSID |
+| Decap mapper | SAI_TUNNEL_MAP_TYPE_VSID_TO_BRIDGE_IF |
+
+### CLI/YANG model Enhancements 
+
+#### NVGRE Tunnel YANG model
+
+```yang
+module sonic-nvgre-tunnel {
+
+    yang-version 1.1;
+
+    namespace "http://github.com/Azure/sonic-nvgre-tunnel";
+    prefix nvgre;
+
+    import ietf-inet-types {
+        prefix inet;
+    }
+
+    import sonic-vlan {
+        prefix vlan;
+    }
+
+    description "NVGRE Tunnel YANG Module for SONiC OS";
+
+    revision 2021-10-31 {
+        description
+            "First Revision";
+    }
+
+    container sonic-nvgre-tunnel {
+
+        container NVGRE_TUNNEL {
+
+            description "NVGRE_TUNNEL part of config_db.json";
+
+            list NVGRE_TUNNEL_LIST {
+
+                key "tunnel_name";
+
+                leaf tunnel_name {
+                    description "NVGRE Tunnel name";
+
+                    type string {
+                        length 1..255;
+                    }
+                }
+
+                leaf src_ip {
+                    description "Source IP address";
+
+                    mandatory true;
+                    type inet:ip-address;
+                }
+
+            }
+            /* end of NVGRE_TUNNEL_LIST */
+
+        }
+        /* end of container NVGRE_TUNNEL */
+
+        container NVGRE_TUNNEL_MAP {
+
+            description "NVGRE_TUNNEL_MAP part of config_db.json";
+
+            list NVGRE_TUNNEL_MAP_LIST {
+
+                key "tunnel_name vlan_name";
+
+                leaf tunnel_name {
+                    description "NVGRE Tunnel name";
+
+                    type leafref {
+                        path /nvgre:sonic-nvgre-tunnel/nvgre:NVGRE_TUNNEL/nvgre:NVGRE_TUNNEL_LIST/nvgre:tunnel_name;
+                    }
+                }
+
+                leaf vlan_name {
+                    description "VLAN name";
+
+                    type leafref {
+                        path /vlan:sonic-vlan/vlan:VLAN/vlan:VLAN_LIST/vlan:name;
+                    }
+                }
+
+                leaf vlan_id {
```

Comment:
why do we need both vlan name and vlan id? having just one is not enough?

Reply:
Because the `vlan_name` is a part of the `NVGRE_TUNNEL_MAP` key, but `vlan_id`it is a number that is used for mapping between the `vsid` number.

Reply:
vlan_name and vlan-id are duplicate information, please try to use one, maybe use vlan_id as key instead of vlan_name.

Reply:
This `Config DB` node design - `vlan_name` and `vlan_id` was added to be consistent with the existing `VxLAN Tunnel` feature.
Also, the user can set to `vlan_name` anything he wants, because it is part of the key (`NVGRE_TUNNEL_MAP`), but in the `orchagent` implementation I'm using `vlan_id` because it is consistent with the `SAI` model.